### PR TITLE
RATIS-1983. Refactor client request processing to support reference count.

### PR DESCRIPTION
**Pull request by szetszwo**
See RATIS-1983

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java [hunk 1]
```diff
@@ -246,13 +247,48 @@ public LogEntryProto load(LogRecord key) throws IOException {
...
+
+    void put(TermIndex key, ReferenceCountedObject<LogEntryProto> valueRef, Op op) {
+      valueRef.retain();
+      Optional.ofNullable(map.put(key, valueRef)).ifPresent(this::release);
+      size.getAndAdd(getEntrySize(valueRef.get(), op));
+    }
+
+    private void release(ReferenceCountedObject<LogEntryProto> entry) {
+      size.getAndAdd(-getEntrySize(entry.get(), Op.REMOVE_CACHE));
+      entry.release();
```
Comment (duongkame) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java:278]: `getEntrySize` may result in different results based on `op`. If `put` and `release` use different `op`, may the `size` calculation be wrong?

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java [hunk 2]
```diff
@@ -202,8 +204,9 @@ public void testAppendEntries() throws Exception {
     while (size < max) {
       SimpleOperation op = new SimpleOperation("m" + i);
       LogEntryProto entry = LogProtoUtils.toLogEntryProto(op.getLogEntryContent(), term, i++ + start);
-      size += getEntrySize(entry, LogSegment.Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
-      segment.appendToOpenSegment(entry, LogSegment.Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
+      size += getEntrySize(entry, Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
+      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry)
+      );
```
Comment (duongkame) [inline ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java:209]:
```suggestion
      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry));
```

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java [hunk 3]
```diff
@@ -235,18 +238,21 @@ public void testAppendWithGap() throws Exception {
     final StateMachineLogEntryProto m = op.getLogEntryContent();
     try {
       LogEntryProto entry = LogProtoUtils.toLogEntryProto(m, 0, 1001);
-      segment.appendToOpenSegment(entry, LogSegment.Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
+      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry)
+      );
```
Comment (duongkame) [inline ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java:242]:
```suggestion
      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry));
```

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java [hunk 4]
```diff
@@ -235,18 +238,21 @@ public void testAppendWithGap() throws Exception {
...
       Assert.fail("should fail since the entry's index needs to be 1000");
     } catch (IllegalStateException e) {
       // the exception is expected.
     }
 
     LogEntryProto entry = LogProtoUtils.toLogEntryProto(m, 0, 1000);
-    segment.appendToOpenSegment(entry, LogSegment.Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
+    segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry)
+    );
```
Comment (duongkame) [inline ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java:250]:
```suggestion
    segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry));
```

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java [hunk 5]
```diff
@@ -235,18 +238,21 @@ public void testAppendWithGap() throws Exception {
...
 
     try {
       entry = LogProtoUtils.toLogEntryProto(m, 0, 1002);
-      segment.appendToOpenSegment(entry, LogSegment.Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
+      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry)
+      );
```
Comment (duongkame) [inline ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java:255]:
```suggestion
      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry));
```

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java [hunk 6]
```diff
@@ -261,7 +267,8 @@ public void testTruncate() throws Exception {
     for (int i = 0; i < 100; i++) {
       LogEntryProto entry = LogProtoUtils.toLogEntryProto(
           new SimpleOperation("m" + i).getLogEntryContent(), term, i + start);
-      segment.appendToOpenSegment(entry, LogSegment.Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE);
+      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry)
+      );
```
Comment (duongkame) [inline ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestLogSegment.java:271]:
```suggestion
      segment.appendToOpenSegment(Op.WRITE_CACHE_WITHOUT_STATE_MACHINE_CACHE, ReferenceCountedObject.wrap(entry));
```

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java [hunk 7]
```diff
@@ -421,14 +423,17 @@ protected CompletableFuture<Long> appendEntryImpl(LogEntryProto entry, Transacti
       // If the entry has state machine data, then the entry should be inserted
       // to statemachine first and then to the cache. Not following the order
       // will leave a spurious entry in the cache.
-      CompletableFuture<Long> writeFuture =
-          fileLogWorker.writeLogEntry(entry, context).getFuture();
+      final CompletableFuture<Long> writeFuture = fileLogWorker.writeLogEntry(entry, context).getFuture();
+      final Function<LogEntryProto, ReferenceCountedObject<LogEntryProto>> wrap = context != null?
```
Comment (duongkame) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java:427]:
```suggestion
      final Function<LogEntryProto, ReferenceCountedObject<LogEntryProto>> wrap = context != null ?
```

**Review by duongkame — APPROVED** [review]
Thanks for the patch @szetszwo . This is very helpful and I learned a lot from it.
LGTM. I just have put a few minor comments.
Also, I think we can have some simple unit tests, e.g. in `TestLogSegment`, to verify proper `release` after log entries are evicted.

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java (hunk 1 above)
Comment (szetszwo) [reply ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java:278]: You are right that `getEntrySize(..)` seems returning different sizes for different op, although the size returned are always the same in the current usage.  Let me refactor it.

**Comment by szetszwo** [comment]
@duongkame , thanks for reviewing this!  Just have pushed a commit addressing your comments.

**Review by SzyWilliam — APPROVED** [review]
+1 the change looks good! 
I've learned a lot on/from how zero copy is implemented. Keenly looking forward to use it in the next release.

**Comment by szetszwo** [comment]
@duongkame , @SzyWilliam , thanks a lot for reviewing this!

**Pull request merged**